sample-app/src/utils/geolocationutils.tsx

```diff
@@ -0,0 +1,56 @@
+import { AnswersActions } from "@yext/answers-headless-react";
+import { SearchIntent } from "@yext/answers-headless";
+
+const defaultGeolocationOptions: PositionOptions = {
+  enableHighAccuracy: false,
+  timeout: 6000,
+  maximumAge: 300000,
+};
+
+/**
+ * Executes universal or vertical search with user location retrieved from nagivator.geolocation API.
+ * If useLocationOnNearMeOnly is true, will attempt to get user location only if the query's searchIntents
+ * (retrieved through autocomplete response) contains 'NEAR_ME'. If there's a failure in getting
+ * user location, a normal search is performed.
+ */
+export async function executeSearchWithUserLocation(
+  answersActions: AnswersActions,
+  isVertical: boolean,
+  geolocationOptions?: PositionOptions,
+  useLocationOnNearMeOnly?: boolean
+) {
+  const executeSearch = () => {
+    isVertical
+      ? answersActions.executeVerticalQuery()
+      : answersActions.executeUniversalQuery();
+    return;
+  }
+
+  if (useLocationOnNearMeOnly) {
+    const results = isVertical
+      ? await answersActions.executeVerticalAutocomplete()
+      : await answersActions.executeUniversalAutocomplete();
+    if (!results?.inputIntents?.includes(SearchIntent.NearMe)) {
+      executeSearch();
+      return;
+    }
+  }
+
+  if ('geolocation' in navigator) {
+    navigator.geolocation.getCurrentPosition((position: GeolocationPosition) => {
+      answersActions.setUserLocation({
+        latitude: position.coords.latitude,
+        longitude: position.coords.longitude,
+      })
+    },
+    (err) => {
+      console.error(err);
+      console.error('Unable to determine user\'s location.');
+    },
+    Object.assign(defaultGeolocationOptions, geolocationOptions)
+    );
+  } else {
+    console.warn('Unable to determine user\'s location.');
+  }
```

Comment:
This function is doing a lot. It might be worth considering turning this into some sore of class, such as a `SearchHandler`. We may want to use a behavioral design pattern here. I'm trying to think which one might be most appropriate.

Reply:
Specifically, I think we should try and decompose what's going on in here. There are multiple responsibilities in this method: fire off a search, see if there's near me intent, and get the user's location. This might be total overkill, but we could use the `Command` pattern and generate three commands:
- SearchCommand
- FetchLocationCommand
- FetchSearchIntentsCommand

Those aren't necessarily the best names. Just an idea, but I think we should consider changing this up to be more than just a single complex method.

Reply:
I briefly tried separating it out before this but thought it might be a little overkill too since the components would have to manage the async stuff and piece together multiple functions . But yeah I can change so it's not a single method

Reply:
updated to have it as a class with static functions